File: src/main/java/org/goafabric/invoice/process/steps/AuthorizationStep.java
package org.goafabric.invoice.process.steps;

import org.goafabric.invoice.controller.extensions.TenantContext;
import org.goafabric.invoice.process.adapter.authorization.Lock;
import org.goafabric.invoice.process.adapter.authorization.LockAdapter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

@Component
public class AuthorizationStep {
    private final Logger log = LoggerFactory.getLogger(this.getClass());

    private final LockAdapter lockAdapter;


    public AuthorizationStep(LockAdapter lockAdapter) {
        this.lockAdapter = lockAdapter;
    }

    public Lock acquireLock() {
        log.info("acquire lock");
        var lock = lockAdapter.acquireLockByKey("invoice-" + TenantContext.getTenantId());
        if (lock.isLocked()) { throw new IllegalStateException("process is already locked"); } else { return lock; }
    }

    public void releaseLock(Lock lock) {
        if (lock != null) {
            log.info("release lock");
            lockAdapter.removeLockById(lock.id());
        }
    }

}

File: src/main/kotlin/org/goafabric/invoice/process/steps/AuthorizationStep.kt
package org.goafabric.invoice.process.steps

import org.goafabric.invoice.controller.extensions.TenantContext
import org.goafabric.invoice.process.adapter.authorization.LockAdapter
import org.goafabric.invoice.process.adapter.authorization.PermissionAdapter
import org.goafabric.invoice.process.adapter.authorization.dto.Lock
import org.goafabric.invoice.process.adapter.authorization.dto.PermissionCategory
import org.goafabric.invoice.process.adapter.authorization.dto.PermissionType
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

@Component
class AuthorizationStep(private val lockAdapter: LockAdapter, private val permissionAdapter: PermissionAdapter) {
    private val log: Logger = LoggerFactory.getLogger(this.javaClass)

    fun acquireLock(): Lock {
        log.info("acquire lock")
        val lock = lockAdapter.acquireLockByKey("invoice-" + TenantContext.getTenantId())
        check(!lock.isLocked) { "process is already locked" }
        return lock
    }

    fun releaseLock(lock: Lock?) {
        if (lock != null) {
            log.info("release lock")
            lockAdapter.removeLockById(lock.id)
        }
    }

    fun checkAuthorization() {
        log.info("check authorization")
        check(
            permissionAdapter.hasPermission(
                TenantContext.getUserName(),
                PermissionCategory.PROCESS,
                PermissionType.INVOICE
            )
        ) { "User " + TenantContext.getUserName() + " is not allowed to execute process" }
    }
}
